src/framework/uicomponents/view/validators/doubleinputvalidator.cpp

```diff
@@ -31,67 +31,86 @@ DoubleInputValidator::DoubleInputValidator(QObject* parent)
 
 void DoubleInputValidator::fixup(QString& string) const
 {
-    auto zeros = [](int num)->QString {
-        QString s = QString();
-        for (int i = 0; i < num; i++) {
-            s.append("0");
+    auto removeTrailingZeros = [](QString str)->QString {
+        QString s = str;
+
+        if (!s.contains('.')) {
+            return s;
+        }
+
+        size_t num = s.size();
+        for (size_t i = num - 1; i > 0; i--) {
+            if (s[i] == '0') {
+                s.remove(i, 1);
+            } else if (s[i] == '.') {
+                s.remove(i, 1);
+                break;
+            } else {
+                break;
+            }
         }
         return s;
     };
 
-    if (!string.contains(".")) {
-        string.append("." + zeros(m_decimal));
-    }
-
     if (string.startsWith(".")) {
         string.prepend("0");
     }
 
     if (string.endsWith(".")) {
-        string.append(zeros(m_decimal));
+        string.remove(string.size() - 1, 1);
     }
 
     QStringList strList = string.split(".", Qt::SkipEmptyParts);
 
     QString intPart = strList.at(0);
-    QString floatPart = strList.at(1);
-
-    if (floatPart.length() < m_decimal) {
-        floatPart.append(zeros(m_decimal - floatPart.length()));
-    }
+    QString floatPart = strList.size() > 1 ? strList.at(1) : 0;
 
     if (intPart.contains(QRegularExpression("^0{1,3}$"))) {
         intPart = QString("0");
     } else if (intPart.contains(QRegularExpression("^\\-0{0,3}$"))) {
         intPart = QString("-0");
     }
 
-    if (intPart == QString("-0") && floatPart == zeros(m_decimal)) {
+    if (intPart == QString("-0") && floatPart.isEmpty()) {
         intPart = QString("0");
     }
 
+    if (floatPart.size() > m_decimal) {
+        floatPart = floatPart.remove(m_decimal, floatPart.size() - m_decimal);
+    }
+
     string = QString("%1.%2").arg(intPart).arg(floatPart);
 
     if (string.toDouble() > m_top) {
         string = QString::number(m_top, 'f', m_decimal);
     } else if (string.toDouble() < m_bottom) {
         string = QString::number(m_bottom, 'f', m_decimal);
     }
+
+    string = removeTrailingZeros(string);
 }
 
 QValidator::State DoubleInputValidator::validate(QString& inputStr, int& cursorPos) const
 {
     QValidator::State state = Invalid;
 
-    if (inputStr.contains(QRegularExpression(QString("^\\-?\\d{1,3}\\.\\d{%1}$").arg(m_decimal)))) {
-        if (inputStr.contains(QRegularExpression("^\\-?0{2,3}\\."))
-            || (inputStr.startsWith("-") && muse::RealIsNull(inputStr.toDouble()))) {
+    static const QRegularExpression validRegex(QString("^\\-?\\d{1,3}(\\.\\d{1,%1})?$").arg(m_decimal));
```

Comment:
It looks like it won't work correctly for different values of m_decimal. We will construct it only once and it will never get updated